onnxruntime/core/providers/cpu/tensor/scatter.cc

```diff
@@ -189,17 +233,17 @@ Status CopyScatterData(const FuncT& func, const Tensor* data_input, const Tensor
   return Status::OK();
 }
 
-template <class T, class... Args>
-inline Status CopyInt32Index(Args&&... args) {
-  return CopyScatterData<int32_t, T>(Func_Assignment<T>(), std::forward<Args>(args)...);
-}
-
-template <class T, class... Args>
-inline Status CopyInt64Index(Args&&... args) {
-  return CopyScatterData<int64_t, T>(Func_Assignment<T>(), std::forward<Args>(args)...);
-}
+template <typename TData>
+struct CopyScatterDataDispatchTarget {
+  Status operator()(const Tensor* data_input, const std::vector<int64_t>& indices_data, const Tensor* updates_input, int64_t axis,
+                    Tensor* data_output) const {
+    return CopyScatterData<TData>(
+        Func_Assignment<TData>(), data_input, indices_data, updates_input, axis, data_output);
+  }
+};
 
-Status Scatter::Compute(OpKernelContext* context) const {
+template <typename EnabledDataTypes>
+Status Scatter<EnabledDataTypes>::Compute(OpKernelContext* context) const {
```

Comment:
Do we need to templatize the whole thing, or could we have a base class that take a flag saying which set of types to use?

Reply:
hm, not sure I follow. the flag would need to be known at compile time, right? something like `Base<bool IsScatterElements>` would still require two separate types. let me know if I'm missing something. we can refine this further in another PR.

---
In reply to: [603774084](https://github.com/microsoft/onnxruntime/pull/7171#discussion_r603774084) [](ancestors = 603774084)

Reply:
just for reference, this could be done with a virtual method or base class flag to indicate whether the kernel is ScatterElements (or Scatter). then we have a condition in Compute() which checks the flag and uses the appropriate set of enabled types.